src/core/autodiff.cc

```diff
@@ -0,0 +1,387 @@
+        required_index_vars_)
+      throw std::runtime_error(
+          "Not all index variables are used in gradient comprehension. "
+          "AD will require range inference to support this case.");
+  }
+
+  bool hasZeroGrad(const std::string& name) {
+    return has_zero_grad_.count(name) > 0;
+  }
+  void markZeroGrad(const std::string& name) {
+    has_zero_grad_.count(name);
+  }
+
+  std::vector<lang::TreeRef>&& getGradComps() {
+    return std::move(grad_comps_);
+  }
+
+  void requireAllIndexVarsOf(const Comprehension& comp) {
+    required_index_vars_ = usedIndexVars(comp);
+  }
+
+  Ident gradNameOf(const Ident& primal_name) {
+    if (has_writeable_grad_.count(primal_name.name()) > 0) {
+      return makeGradName(primal_name);
+    }
+    return seedNameOf(primal_name);
+  }
+  Ident seedNameOf(const Ident& primal_name) {
+    return makeSeedName(makeGradName(primal_name));
+  }
+
+ private:
+  Ident makeGradName(const Ident& name) {
+    return Ident(Ident::create(dummyRange, std::string("d_") + name.name()));
+  }
+
+  Ident makeSeedName(const Ident& name) {
+    return Ident(Ident::create(dummyRange, std::string("seed_") + name.name()));
+  }
+
+  std::unordered_set<std::string> required_index_vars_;
+  std::vector<lang::TreeRef> grad_comps_;
+  // Keys in these sets are always names of primal variables.
+  std::unordered_set<std::string> primal_outputs_;
+  std::unordered_set<std::string> has_writeable_grad_;
+  std::unordered_set<std::string> has_zero_grad_;
+};
+
+void differentiateExpr(
+    GradInfo& grad_info,
+    lang::TreeRef expr,
+    lang::TreeRef grad_output_expr) {
+  using namespace lang;
+  switch (expr->kind()) {
+    case TK_ACCESS: {
+      Access acc{expr};
+      grad_info.addGradComprehension(
+          acc.name(), acc.arguments(), grad_output_expr);
+      break;
+    }
+    case '+': {
+      differentiateExpr(grad_info, expr->tree(0), grad_output_expr);
+      differentiateExpr(grad_info, expr->tree(1), grad_output_expr);
+      break;
+    }
+    case '*': {
+      differentiateExpr(
+          grad_info,
+          expr->tree(0),
+          Compound::create(
+              '*', expr->range(), {grad_output_expr, expr->tree(1)}));
+      differentiateExpr(
+          grad_info,
+          expr->tree(1),
+          Compound::create(
+              '*', expr->range(), {grad_output_expr, expr->tree(0)}));
+      break;
+    }
+    case TK_CONST: {
+      // There's nothing we have to do to handle constants, because we don't
+      // differentiate w.r.t. them.
+      break;
+    }
+    default:
+      throw ErrorReport(expr) << "Unsupported expression kind in AD: "
+                              << kindToString(expr->kind());
+  }
+}
+
+// XXX: Sema isn't nilpotent, so we have to reparse the source
+std::vector<TreeRef> inferOutputTypes(const std::string& source) {
+  auto halide_def =
+      tc2halide::translate(isl::with_exceptions::globalIslCtx(), source, true);
```

Comment:
So this now makes lang depend on ISL which is unnecessary.
We should extract what we need from halide for range inference.